### PR TITLE
Integrate Cloud storage web extension with shield utils

### PR DESCRIPTION
**Pull request by punamdahiya**

### CODE REVIEW

**Comment by punamdahiya** [comment]
@motin @gregglind I have taken a first pass at integrating shield-utils with cloud storage web extension. This has telemetry probes missing which I will take a stab at next. Wanted to share with you for early feedback. Thanks!

**Comment by punamdahiya** [comment]
@gregglind @motin last commit in this PR integrate telemetry, https://github.com/mozilla/shield-cloudstorage/pull/28/files#diff-7f87af4f00f7747a008141453c7eb047R91 in background.js is getting respective probes data values.

PR can be picked and validated as part of fixing https://github.com/mozilla/shield-studies-addon-utils/issues/173. If possible, it will help when ready, we can submit updated v5 shield utils PR against this PR. Thanks!

**Comment by punamdahiya** [comment]
@gijs including you to review integrating telemetry probes part of the extension. Thanks!

File: webextension/study.js [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+  const id = browser.runtime.id;
+  const prefs = {
+    variation: `shield.${id}.variation`,
+    firstRunTimestamp: `shield.${id}.firstRunTimestamp`,
+  };
+  prefs;
+  studySetup.allowEnroll = await shouldAllowEnroll();
+  studySetup.testing = {
+    // variation: await browser.prefs.getStringPref(prefs.variation);
+    // firstRunTimestamp: await browser.prefs.getStringPref(prefs.firstRunTimestamp);
```
Comment (gijsk) [inline webextension/study.js]: What is this, and if this is commented out does that mean that we basically don't need the `browser.prefs` implementation in this PR - I don't see any other accesses?

File: webextension/study.js [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+/**
+ * Augment studySetup with a few async values
+ */
+async function getStudySetup() {
+  const id = browser.runtime.id;
+  const prefs = {
+    variation: `shield.${id}.variation`,
+    firstRunTimestamp: `shield.${id}.firstRunTimestamp`,
+  };
+  prefs;
```
Comment (gijsk) [inline webextension/study.js]: This line doesn't do anything.

File: webextension/privileged/study/api.js [hunk 3]
```diff
@@ -0,0 +1,2055 @@
...
...u+" "+g+"= "+a+" : "+u+" "+g+" "+a+") || "+u+" !== "+u+") { var op"+o+" = exclusive"+o+" ? '"+y+"' : '"+y+"=';"}else{var E=!0===m,w=y;E||(w+="=");var b="'"+w+"'";s+=" if ( ",f&&(s+=" ("+a+" !== undefined && typeof "+a+" != 'number') || "),s+=" "+u+" "+g,E&&(s+="="),s+=" "+a+" || "+u+" !== "+u+") {"}var t=r,j=j||[];j.push(s),s="",!1!==e.createErrors?(s+=" { keyword: '"+(t||"_limit")+"' , dataPath: (dataPath || '') + "+e.errorPath+" , schemaPath: "+e.util.toQuotedString(c)+" , params: { comparison: "+b+", limit: "+a+", exclusive: "+E+" } ",!1!==e.opts.messages&&(s+=" , message: 'should be "+w+" ",s+=f?"' + "+a:n+"'"),e.opts.verbose&&(s+=" , schema:  ",s+=f?"validate.schema"+l:""+n,s+="         , parentSchema: validate.schema"+e.schemaPath+" , data: "+u+" "),s+=" } "):s+=" {} ";var S=s;return s=j.pop(),s+=!e.compositeRule&&h?e.async?" throw new ValidationError(["+S+"]); ":" validate.errors = ["+S+"]; return false; ":" var err = "+S+";  if (vErrors === null) vErrors = [err]; else vErrors.push(err); errors++; ",s+=" } ",h&&(s+=" else { "),s}},{}],15:[function(e,r,t){"use strict";r.exports=function(e,r){var t,a,s=" ",o=e.level,i=e.dataLevel,n=e.schema[r],l=e.schemaPath+e.util.getProperty
```
Comment (gijsk) [inline webextension/privileged/study/api.js]: We should probably not ship with unconditional `console.log` statements.

File: webextension/privileged/ui/CloudDownloadsView.jsm [hunk 4]
```diff
@@ -230,16 +227,46 @@ var CloudDownloadsView = {
...
         }
         return;
       }
-      await this.initWindowListener();
+
+      await this.handleAddOnInitTelemetry();
+      if (this.studyVariation !== "control") {
+        await this.initWindowListener();
+      }
       this.registerNotification();
```
Comment (gijsk) [inline webextension/privileged/ui/CloudDownloadsView.jsm]: Why do we need to do the non-window-listener thing even in the control case?

File: webextension/study.js [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+  // Cached answer.  Used on 2nd run
+  let allowed = await browser.storage.local.get("allowedToEnroll");
+  if (allowed) return true;
+
+  /*
+  First run, we must calculate the answer.
+  If false, the study will endStudy with 'ineligible' during `setup`
+  */
+  // could have other reasons to be eligible, such add-ons, prefs
+  const dataPermissions = await browser.study.dataPermissions();
```
Comment (gijsk) [inline webextension/study.js]: What implements this? I don't see anything that does, except maybe `getDataPermissions`, in which case this call is incorrect...

**Comment by gijsk** [comment]
> @gijs including you to review integrating telemetry probes part of the extension. Thanks!

Can you provide some context to what's going on in the PR? Like, why is the initialization of the providers moving? What bits am I actually supposed to review - what's the giant webpack `study/api.js` file, and was it already reviewed? How was it built - I assume it's easier to review the original rather than the webpacked thing in this PR?

This is very much not "just" some telemetry (which is usually just a few lines + the schema definition of the telemetry data itself) - it changes how the add-on initializes, adds the concept of 'variations', and so on - but the docs modifications are minimal and there are no details in this PR, so it's difficult to understand what's going on.

**Comment by gijs** [comment]
I keep getting notifications but I'm another Gijs :)

**Comment by gijsk** [comment]
Sorry! @punamdahiya , you probably want @gijsk in future.

**Comment by punamdahiya** [comment]
Oops my bad! will correct going forward.

**Comment by punamdahiya** [comment]
@gijsk Sorry for confusion, PR has two commits. I should have pointed to below exact commit (Integrate Telemetry probes) to be reviewed

https://github.com/mozilla/shield-cloudstorage/pull/28/commits/2d9d464fc91c5fbdbb631ca8c3fdaf7c705cd5c9

The first commit in the PR ‘Integrate Shield Utils with Cloud Storage extension’ will be [updated by](https://github.com/punamdahiya/shield-cloudstorage/blob/08d35d846d71c9ee8b98b8a26680bcdbfb0589ce/README.md) reviewed work on the v5 shield utils API (It’s currently in QA , link below points to development repo).

https://github.com/mozilla/shield-studies-addon-utils/tree/develop

npm install --save mozilla/shield-studies-addon-utils#develop
Once we have v5 shield utils ready, the integrating shield util part can be reviewed by @gregglind and @motin

File: webextension/privileged/ui/CloudDownloadsView.jsm (hunk 4 above)
Comment (punamdahiya) [reply webextension/privileged/ui/CloudDownloadsView.jsm]: Control branch will not have any of cloud storage feature UI  which is setup inside initWindowListener.  To compare  if there is difference in downloads with or without cloud storage feature, for all branches we are logging
a) how many download triggered by the user (download_started)
https://github.com/mozilla/shield-cloudstorage/pull/28/commits/2d9d464fc91c5fbdbb631ca8c3fdaf7c705cd5c9#diff-7e1f5d4fd67bdee7405adbadd873d06bR279

b) Number of cloud storage providers (addon_init inside handleAddOnInitTelemetry)
https://github.com/mozilla/shield-cloudstorage/pull/28/commits/2d9d464fc91c5fbdbb631ca8c3fdaf7c705cd5c9#diff-7e1f5d4fd67bdee7405adbadd873d06bR266

Rest of the telemetry documented inside /docs/telemetry.json, will be only for experimental branches showing cloud storage feature UI

Will update docs with branches tested in the study

File: webextension/privileged/study/api.js (hunk 3 above)
Comment (punamdahiya) [reply webextension/privileged/study/api.js]: Part of first commit (See comment below)

File: webextension/study.js (hunk 5 above)
Comment (punamdahiya) [reply webextension/study.js]: Part of first commit (See comment below)

File: webextension/privileged/ui/CloudDownloadsView.jsm (hunk 4 above)
Comment (punamdahiya) [reply webextension/privileged/ui/CloudDownloadsView.jsm]: @gijsk About moving initialization of providers inside handleAddOnInitTelemetry, we are doing so to invoke this method for all branches including control.

Moving it outside of initWindowListener gives us an additional benefit to initialize providers only one time if multiple windows are open.

**Comment by punamdahiya** [comment]
Updated https://github.com/mozilla/shield-cloudstorage/pull/28/commits/db0d5f4f747ea6c6ad32d817718e0643f4f04888 with post survey URL and variation section in ReadMe

File: webextension/privileged/ui/CloudDownloadsView.jsm [hunk 6]
```diff
@@ -533,6 +565,13 @@ var CloudDownloadsView = {
         CloudDownloadsInternal.selectedProvider = { providerKey, value: this.providers.get(providerKey) };
         CloudDownloadsInternal.handleMove(download);
       }
+      telemetryData = {
+        message: "prompt_move_download_context_menu",
+        provider: providerKey,
+        provider_count: this.providers.size.toString(),
+        timestamp: Math.floor(Date.now() / 1000).toString(),
```
Comment (gijsk) [inline webextension/privileged/ui/CloudDownloadsView.jsm]: Why do you need a timestamp that's this granular?

File: docs/telemetry.json [hunk 7]
```diff
@@ -8,7 +8,7 @@
   "definitions": {
     "branch_message": {
       "type": "string",
-      "enum": ["addon_init", "download_prefs", "prompt_save_click", "prompt_cancel_click",
+      "enum": ["addon_init", "download_prefs", "prompt_opted_in", "prompt_cancel_click",
```
Comment (gijsk) [inline docs/telemetry.json]: This doesn't include the install/uninstall messages.

File: webextension/privileged/ui/CloudDownloadsView.jsm [hunk 8]
```diff
@@ -230,16 +227,46 @@ var CloudDownloadsView = {
...
 
+  // Log provider count and provider keys existing on user desktop
+  async handleAddOnInitTelemetry() {
+    // There is no other API that awaits the result of CloudStorageInternal.init()
+    // Invoke getDownloadFolder on CloudStorage API to ensure API is initialized
+    // before asking CloudStorage API for storage providers info
+    await CloudStorage.getDownloadFolder();
+    const providers = await CloudStorage.getStorageProviders();
+
+    const keys = [];
```
Comment (gijsk) [inline webextension/privileged/ui/CloudDownloadsView.jsm]:
```js
const keys = Array.from(providers.keys());
```

File: webextension/privileged/ui/CloudDownloadsView.jsm [hunk 9]
```diff
@@ -545,19 +584,42 @@ var CloudDownloadsView = {
         // Prepare for future downloads move to Download Folder by checking
         // if Download folder exists in cloud provider folder, if not create one
         CloudDownloadsInternal.checkProviderDownloadFolder();
+
+        // Once opted-in, set observers to send subsequent download pref changes to telemetry
+        Services.prefs.addObserver("browser.download.folderList", CloudDownloadsInternal.downloadPrefObserve);
+        Services.prefs.addObserver("browser.download.useDownloadDir", CloudDownloadsInternal.downloadPrefObserve);
```
Comment (gijsk) [inline webextension/privileged/ui/CloudDownloadsView.jsm]: You never remove these preference observers.

It's probably easier to just piggyback on the existing lazy preference getter we have for this, and check in the pref update callback whether the user has opted in (ie whether we've saved a prompt response). You'd need to add another lazy pref getter for the other pref though.

File: webextension/background.js [hunk 10]
```diff
@@ -86,7 +86,14 @@ class StudyLifeCycleHandler {
         break;
       }
 
-      browser.cloudstorage.init(browser.runtime.getURL("./skin/clouddownloads.css"), interval).then(
+      browser.cloudstorage.onRecordTelemetry.addListener(
```
Comment (gijsk) [inline webextension/background.js]: Presumably we need to remove this listener at some point, too?

File: webextension/background.js [hunk 11]
```diff
@@ -86,7 +86,14 @@ class StudyLifeCycleHandler {
...
+        (value) => {
+          console.log("onRecordTelemetry", value);
```
Comment (gijsk) [inline webextension/background.js]: We really shouldn't ship with `console.log` still in the code...

File: webextension/privileged/ui/CloudDownloadsView.jsm (hunk 6 above)
Comment (punamdahiya) [reply webextension/privileged/ui/CloudDownloadsView.jsm]: Started storing last prompt time inside cloud.services.lastPrompt pref in seconds and to keep the format consistent, kept  time in seconds across Cloud Storage

File: docs/telemetry.json (hunk 7 above)
Comment (punamdahiya) [reply docs/telemetry.json]: Install and uninstall telemetry should be coming from shield utils. Updated to remove cloud_storage_install and cloud_storage_uninstall telemetry. 

Updated doc with a new telemetry  'existing_cloud_provider_download_settings', to log users that have cloud provider as their default downloads settings before opt-in from notification panel. These users are not shown notification and Move to <cloud provider> option in context menu.

File: webextension/privileged/ui/CloudDownloadsView.jsm (hunk 8 above)
Comment (punamdahiya) [reply webextension/privileged/ui/CloudDownloadsView.jsm]: +1

File: webextension/privileged/ui/CloudDownloadsView.jsm (hunk 9 above)
Comment (punamdahiya) [reply webextension/privileged/ui/CloudDownloadsView.jsm]: Done

File: webextension/background.js (hunk 11 above)
Comment (punamdahiya) [reply webextension/background.js]: +1, removed console here. Couple that are still there are during initialization in background.js, will be updating those with shield utils integration.

File: webextension/background.js (hunk 10 above)
Comment (punamdahiya) [reply webextension/background.js]: My understanding that's taken care by EventEmitter  here https://github.com/mozilla/shield-cloudstorage/pull/28/commits/d785003c2c1009f50060646f591a0b59c6fc4942#diff-549919a14832a5cacd7ae883b94a4db1R78

**Comment by punamdahiya** [comment]
@gijsk Updated PR with review feedback. I have left review feedback updates as separate commit for now. Will squash before merging with master. Please review. Thanks!

File: webextension/privileged/ui/CloudDownloadsView.jsm (hunk 6 above)
Comment (gijsk) [reply webextension/privileged/ui/CloudDownloadsView.jsm]: OK, but then, why do you need a timestamp in the telemetry at all? And why couldn't it just be, say, the date?

Getting second-granularity timestamps feels like something that has privacy implications. Please think carefully about whether we actually need that for the study we're doing.

**Comment by gijsk** [comment]
Please see earlier comment about the timestamps, and make sure all the `console.*` calls make their way out of the code before shipping.

File: webextension/privileged/ui/CloudDownloadsView.jsm (hunk 6 above)
Comment (punamdahiya) [reply webextension/privileged/ui/CloudDownloadsView.jsm]: sounds good, updated time stamp to use new Date()

**Comment by punamdahiya** [comment]
@gijsk Thanks for review! Updated console.* statements to use browser.study.log to be exposed via shield utils api. Left commented for now till shield utils api makes these log calls functional.

**Comment by gregglind** [comment]
(peeps:  actual working v5 release candidate for shield-study-addon-utils at:  https://github.com/mozilla/shield-studies-addon-utils/pull/178)

**Comment by gregglind** [comment]
(I decided to kill log, but it sounds like you need it back!  I gave up on reasonable logging.  @gijsk, ideas welcome.  I am simply running out of steam and time on it.)

**Comment by punamdahiya** [comment]
Merging the PR on master for QA to start testing telemetry probes.

@SoftVision-CosminMuntean extension should log respective probes under about:telemetry. Thanks!

**Comment by gijsk** [comment]
So to be clear, I don't know what granularity of timestamps you require, and although I *think* telemetry events does its own timestamping, I could be wrong. My questions about this have a simple aim: make sure we only track what we really need to track. I'm not sure what analysis is being planned, so I can't make definitive statements about what we need - but we should make sure to get what we need and no more, and store it in a way that makes it as easy as possible to analyze the resulting telemetry data. I don't think the switch to `new Date()` was an improvement in that respect.

File: webextension/background.js [hunk 12]
```diff
@@ -1,10 +1,129 @@
...
+
+  /**
+   * do some cleanup / 'feature reset'
+   *
+   * (If you have privileged code, you might need to clean
+   *  that up as well.
+   * See:  https://firefox-source-docs.mozilla.org/toolkit/components/extensions/webextensions/lifecycle.html
+   */
+  async cleanup() {
+    await browser.storage.local.clear();
```
Comment (gregglind) [inline webextension/background.js]: Claim:  i have heard that this this auto clears, by the way.

File: webextension/background.js [hunk 13]
```diff
@@ -1,10 +1,129 @@
...
+        interval = 1;
+        break;
+      case "notification-interval-longer":
+        interval = 2;
+        break;
+      }
+
+      browser.cloudstorage.onRecordTelemetry.addListener(
+        (value) => {
+          browser.study.sendTelemetry(value);
```
Comment (gregglind) [inline webextension/background.js]: Probably a good idea to have a check that value is conformant (your `value` is an Object of key-value pairs where all are strings).  ShieldStudy convention is to call such a thing a `payload`.

File: webextension/background.js [hunk 14]
```diff
@@ -1,10 +1,129 @@
...
+        },
+      );
+
+      browser.cloudstorage.init(browser.runtime.getURL("./skin/clouddownloads.css"), interval, studyInfo.variation.name).then(
+        result => {
+          // browser.study.log(result);
+        });
+    } catch (e) {
+      // browser.study.log(e);
```
Comment (gregglind) [inline webextension/background.js]: Console error here seems okay to me.

Browser study log needs some more thinking.   I am not sure how to think about "levels" as part of `browser.study.log`.

File: webextension/background.js [hunk 15]
```diff
@@ -1,10 +1,129 @@
...
+    }
+  }
+
+  /** handles `study:end` signals
+   *
+   * - opens 'ending' urls (surveys, for example)
+   * - calls cleanup
+   */
+  async handleStudyEnding(ending) {
```
Comment (gregglind) [inline webextension/background.js]: IMPORTANT:  you need a `browser.study.uninstall()` somewhere in this, after you do your work.

File: webextension/manifest.json [hunk 16]
```diff
@@ -10,10 +10,27 @@
   "author": "Mozilla (https://mozilla.org/)",
   "background": {
     "scripts": [
+      "study.js",
       "background.js"
     ]
   },
   "experiment_apis": {
+    "prefs": {
+      "schema": "./privileged/prefs/schema.json",
```
Comment (gregglind) [inline webextension/manifest.json]: You don't need this entry.  We don't have the Prefs api complete.

File: webextension/manifest.json [hunk 17]
```diff
@@ -28,9 +45,9 @@
     "48": "./skin/sync2cloud.svg",
     "96": "./skin/sync2cloud.svg"
   },
-  "license": "MPL-2.0",
   "manifest_version": 2,
   "permissions": [
-    "management"
+    "management",
+    "storage"
```
Comment (gregglind) [inline webextension/manifest.json]: are you actually using storage?    If it's just for the template, rip this out.  SSAU doesn't actually use storage.

File: webextension/study.js [hunk 18]
```diff
@@ -0,0 +1,141 @@
...
```
Comment (gregglind) [inline webextension/study.js]: This is being dropped in 5.0.  Data permissions is complex.  It's a strong claim that Shield itself will remove studies if data permissions change.  Delaying this until 5.1 in a month or so.

File: webextension/privileged/ui/CloudDownloadsView.jsm [hunk 19]
```diff
@@ -533,6 +562,13 @@ var CloudDownloadsView = {
         CloudDownloadsInternal.selectedProvider = { providerKey, value: this.providers.get(providerKey) };
         CloudDownloadsInternal.handleMove(download);
       }
+      telemetryData = {
+        message: "prompt_move_download_context_menu",
+        provider: providerKey,
+        provider_count: this.providers.size.toString(),
+        timestamp: new Date().toString(),
```
[the remark this answers is not in the document]
Comment (punamdahiya) [reply webextension/privileged/ui/CloudDownloadsView.jsm]: @gijsk misunderstood your comment here https://github.com/mozilla/shield-cloudstorage/pull/28#discussion_r189196190 on storing timestamp in milliseconds as having privacy implication, that's why changed it into human readable format.

We do need more than per-day granularity for analysis. Checked with Teon and it's ok to use  creation_date captured by telemetry events
https://sql.telemetry.mozilla.org/queries/53862/source#142596

Will update to remove timestamp collection from inside extension.

[the remark this answers is not in the document]
Comment (punamdahiya) [reply webextension/privileged/ui/CloudDownloadsView.jsm]: Updated  by removing timestamp https://github.com/mozilla/shield-cloudstorage/commit/45b5c81a623073eeaae94c95818dd6aab4207830

File: webextension/study.js (hunk 18 above)
Comment (punamdahiya) [reply webextension/study.js]: Updated in PR https://github.com/mozilla/shield-cloudstorage/pull/38

File: webextension/manifest.json (hunk 17 above)
Comment (punamdahiya) [reply webextension/manifest.json]: Removed

File: webextension/manifest.json (hunk 16 above)
Comment (punamdahiya) [reply webextension/manifest.json]: Updated to remove prefs api https://github.com/mozilla/shield-cloudstorage/pull/38

File: webextension/background.js (hunk 15 above)
Comment (punamdahiya) [reply webextension/background.js]: Done https://github.com/mozilla/shield-cloudstorage/pull/38

File: webextension/background.js (hunk 14 above)
Comment (punamdahiya) [reply webextension/background.js]: leaving console commented for now till we have better handling

File: webextension/background.js (hunk 13 above)
Comment (punamdahiya) [reply webextension/background.js]: Data is getting reflecting correctly in about:telemetry. Updated variable naming to payload to keep it consistent.

**Pull request merged**